runtime/src/global_request_scheduler_mts.c

```diff
@@ -0,0 +1,333 @@
+
+	return val;
+}
+
+
+uint64_t
+global_request_scheduler_mts_guaranteed_peek(void)
+{
+	return priority_queue_peek(global_request_scheduler_mts_guaranteed);
+}
+
+uint64_t
+global_request_scheduler_mts_default_peek(void)
+{
+	return priority_queue_peek(global_request_scheduler_mts_default);
+}
+
+
+/**
+ * Initializes the variant and registers against the polymorphic interface
+ */
+void
+global_request_scheduler_mts_initialize()
+{
+	global_request_scheduler_mts_guaranteed = priority_queue_initialize(RUNTIME_RUNQUEUE_SIZE, false,
+	                                                                            module_request_queue_get_priority);
+	global_request_scheduler_mts_default    = priority_queue_initialize(RUNTIME_RUNQUEUE_SIZE, false,
+                                                                                 module_request_queue_get_priority);
+
+	global_module_timeout_queue = priority_queue_initialize(RUNTIME_RUNQUEUE_SIZE, false,
+	                                                        module_timeout_get_priority);
+
+	LOCK_INIT(&global_lock);
+
+	struct global_request_scheduler_config config = {
+		.add_fn                  = global_request_scheduler_mts_add,
+		.remove_fn               = global_request_scheduler_mts_remove,
+		.remove_if_earlier_fn    = global_request_scheduler_mts_remove_if_earlier,
+		.remove_with_mt_class_fn = global_request_scheduler_mts_remove_with_mt_class,
+		.peek_fn                 = global_request_scheduler_mts_peek
+	};
+
+	global_request_scheduler_initialize(&config);
+}
+
+void
+global_request_scheduler_mts_free()
+{
+	priority_queue_free(global_request_scheduler_mts_guaranteed);
+	priority_queue_free(global_request_scheduler_mts_default);
+	priority_queue_free(global_module_timeout_queue);
+}
+
+
+void
+global_timeout_queue_add(struct module *module)
+{
+	struct module_timeout *mt = malloc(sizeof(struct module_timeout));
+	mt->timeout               = get_next_timeout_of_module(module->replenishment_period, __getcycles());
+	mt->module                = module;
+	priority_queue_enqueue_nolock(global_module_timeout_queue, mt);
+}
+
+/*
+ * Checks if there are any tenant timers that run out in the GLOBAL queue,
+ *  if so promote that tenant.
+ */
+void
+global_timeout_queue_check_for_promotions()
+{
+	uint64_t now = __getcycles();
+
+	struct module_timeout *top_module_timeout = NULL;
+
+	/* Check the timeout queue for a potential tenant to get PRomoted */
+	priority_queue_top_nolock(global_module_timeout_queue, (void **)&top_module_timeout);
+
+	if (top_module_timeout == NULL || now < top_module_timeout->timeout) return;
+
+	struct module *                     module          = top_module_timeout->module;
+	struct module_global_request_queue *mgrq_to_promote = module->mgrq_requests;
+
+	if (/*mgrq_to_promote->mt_class == MT_DEFAULT
+	    && priority_queue_length_nolock(mgrq_to_promote->sandbox_requests) > 0 &&*/
+	    module->remaining_budget <= 0) {
```

Comment:
YOu want to promote periodically. Not just periodically when the budget is expended. Might be missing somethign here.